### PR TITLE
Developer msft

### PR DESCRIPTION
**Pull request by MeritedHobbit**
Version changed to 0.9.0

Below are the changes

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/RAMADH~1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/RAMADH~1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#467886" vlink="#96607D">

Sl No | Change | File Name | Category
-- | -- | -- | --
1 | Addition of Line number in   Redfish command line logs | comptool_dut.py | Framework
2 | Fix   error in some cases when response.dict is used and the response body is empty | comptool_dut.py | Framework
3 | Activate AC   enhancement to consider "Not Implemented" in the o/p of   isGPUReachable() | functional_ifc.py | Interface
4 | R2   Test case check URI before downloading | functional_ifc.py | Interface
5 | Update URI FIX | fw_update_ifc.py | Interface
6 | Supporting   precheck and post check method by getting version details from package json | fw_update_ifc.py | Interface
7 | First we were getting version   from FW inventory, Now we are getting version from bundle package json | fw_update_ifc.py | Interface
8 | Added count for failed component and passed component | fw_update_ifc.py | Interface
9 | Added Tolerance for Version   field & ExpectedVersion to be missing or "" | fw_update_ifc.py | Interface
10 | First we were getting version from FW   inventory, Now we are getting version from bundle package json | fw_update_ifc.py | Interface
11 | Added version_check parameter   (default is true). Needs to be false for select negative scenarios. | fw_update_ifc.py | Interface
12 | Added   check location uri method for validating the uri exists or not before dump   donwload for R2 | ras_ifc.py | Interface
13 | Improved ras ifc file | ras_ifc.py | Interface
14 | Added Pre check into to collect firmware inventory for   negative test cases | CTAMTestNegativeCorruptImageUpdate,   CTAMTestNegativeInvalidDeviceUUIDImageUpdate,   CTAMTestNegativeEmptyMetadataImageUpdate, | TestCase
15 | Added Pre   check into to collect firmware inventory for negative test cases | CTAMTestNegativeInvalidPackageUUIDImageUpdate, | TestCase
16 | Added Pre check into to collect firmware inventory for   negative test cases | CTAMTestNegativeInvalidSignedImageUpdate, | TestCase
17 | Added Pre   check into to collect firmware inventory for negative test cases | CTAMTestNegativeLargeImageUpdate, | TestCase
18 | Added Pre check into to collect firmware inventory for   negative test cases | CTAMTestNegativeUnsignedBundleUpdate, | TestCase
19 | Added Pre   check into to collect firmware inventory for negative test cases | CTAMTestNegativeUnsignedImageUpdate | TestCase
20 | Name   changed for this test case | CTAMTestFullDeviceUpdateStagingInterruptionWithAcReset | TestCase
21 | Added Tag for   CTAMTestFullDeviceUpdateStagingInterruptionWithAcReset | CTAMTestFullDeviceUpdateStagingInterruptionWithAcReset | TestCase
22 | Modified   flow for CTAMTestFullDeviceUpdateStagingInterruptionWithAcReset | CTAMTestFullDeviceUpdateStagingInterruptionWithAcReset | TestCase
23 | Push targets for   CTAMTestSingleDeviceUpdatePingPong | CTAMTestSingleDeviceUpdatePingPong | TestCase
24 | Improved   code for test case CTAMTestServiceValidator | CTAMTestServiceValidator | TestCase
25 | Added method to store config   details before run | test_runner.py | Framework
26 | Added   Progressbar2 for showing progressbar | ctam_utils.py | Framework
27 | Improved code for running   Redfish Service Validator | ctam_utils.ctam_redfish_interop_validator() | Framework
28 | Added   generic method to run all redfish servic3e and interop validator command | ctam_utils.ctam_run_dmtf_command() | Framework
29 | Added fix for key error of   verison | json_utils.jsonmultihunt | Framework
30 | Added   function for jsonmultivaluehunt | jsonutils.jsonmultivaluehunt() | Framework
31 | Moved command line log output   class into logger_utils | Code move from   ctam\testrunner.py to utils\loggerutils | Framework
32 | Fixed   issue with duplicate post processed log files from being created | test_runner.py | Framework

</body>

</html>

Results 

image

### CODE REVIEW

File: ctam/test_runner.py [hunk 1]
```diff
@@ -908,7 +908,103 @@ def generate_domain_test_report(self):
...
+                                        if isinstance(value, list):
+                                            # Convert list to a single line (comma-separated)
+                                            json_data[section_name][key] = ', '.join(str(v) for v in value if v)
+                                        elif isinstance(value, dict):
+                                            for sub_key, sub_value in value.items():
+                                                json_data[section_name][f"{key}_{sub_key}"] = sub_value
+                                        else:
+                                            json_data[section_name][key] = value
+            
+            json_file_path = os.path.join(config_folder_path, "data.json")
```
Comment (dhruvr-nvidia) [inline ctam/test_runner.py:1000]: data.json file needs to be sanitized, depending on the dut_info.json.

Comment (MeritedHobbit) [reply ctam/test_runner.py:1000]: Ohh.....thanks for catching this. Will get back.

File: ctam/interfaces/fw_update_ifc.py [hunk 2]
```diff
@@ -111,37 +111,44 @@ def ctam_fw_update_precheck(self, image_type="default"):
...
+        BundleComponentIdsAndVersions = self.PLDMComponentVersions(image_type=image_type)
         for element in self.PreInstallDetails:
-            # Check ONLY the ones which are part of HttpPushURITargets
+        #     # Check ONLY the ones which are part of HttpPushURITargets
             if (
                 self.included_targets == []
                 or element["@odata.id"] in self.included_targets
             ):
-                msg = f"Pre Install Details: {element['Id']} : {element['SoftwareId']} : {element['Version']} : "
+                msg = f"Pre Install Details: {element['Id']} : {element['SoftwareId']} : {element.get('Version', 'NA')} : "
```
Comment (afsanac-nv) [inline ctam/interfaces/fw_update_ifc.py:123]: Isn't "Version" a mandatory field per OCP spec?

**Review by afsanac-nv — APPROVED** [review]
We've run this branch in Nvidia systems and didn't find any issue. We just had two minor comments as following:

1. Sanitization of data.json file -> We can push the change in our next PR if needed
2. "Version" property mandate -> Can be discussed in weekly meetings

Approving since these aren't blockers. Thanks!

**Review by dhruvr-nvidia — APPROVED** [review]

File: ctam/interfaces/fw_update_ifc.py (hunk 2 above)
Comment (MeritedHobbit) [reply ctam/interfaces/fw_update_ifc.py:123]: Glad you pointed this out :)

There is a discussion, pending conclusion, in OCP FW update spec meeting, where FirmwareInventory was allowed to call out config information. There are still discussing if we should mandate Version field for those cases. Till then, this should help cater to designs which don't have Version. 

Let's revisit this topic for 1.1 spec.

**Pull request merged**